python_modules/dagster-test/dagster_test/toys/asset_sensors.py

```diff
@@ -0,0 +1,119 @@
+from dagster import (
+    AssetKey,
+    AssetOut,
+    RunRequest,
+    SkipReason,
+    asset,
+    job,
+    multi_asset,
+    multi_asset_sensor,
+    op,
+)
+
+
+@asset
+def asset_a():
+    return [1, 2, 3]
+
+
+@asset
+def asset_b():
+    return [5, 6, 7]
+
+
+@asset
+def asset_c():
+    return [11, 12, 13]
+
+
+@asset
+def asset_d():
+    return [15, 16, 17]
+
+
+@multi_asset(
+    outs={
+        "my_string_asset": AssetOut(),
+        "my_int_asset": AssetOut(),
+    }
+)
+def multi_asset_a():
+    return "abc", 123
+
+
+@op
+def log_asset_sensor(context):
+    context.log.info(context.op_config["message"])
+
+
+@job
+def log_asset_sensor_job():
+    log_asset_sensor()
+
+
+@multi_asset_sensor(
+    asset_keys=[AssetKey("asset_a"), AssetKey("asset_b")],
+    job=log_asset_sensor_job,
+)
+def asset_a_and_b_sensor(context, asset_events):
+    asset_events = context.get_latest_event([AssetKey("asset_a"), AssetKey("asset_b")])
+    if all(asset_events.values()):
+        return RunRequest(
+            run_key=context.cursor,
+            run_config={
+                "ops": {
+                    "log_asset_sensor": {"config": {"message": f"Asset events dict {asset_events}"}}
+                }
+            },
+        )
+    else:
+        return None
```

Comment:
Can we discuss a little more around having `return SkipReason(...)` vs `return None` be significant for cursor updating logic?

I have some reservations (especially since this will be new behavior that only applies to this `@multi_asset_sensor`) so want to make sure we explore the full set of potential solutions.

For user-defined sensors, the cursor management logic is maintained by the user themselves, so it is reasonable that they call `context.update_cursor`.  For `@asset_sensor`, we fully manage the cursor management, so the user code doesn't have to do anything.  For these `@multi_asset_sensor`, I can completely see wanting to have some simplified cursor management, but maybe with a cleaner API.

Because we're creating a new context type, I can see us having some custom cursor management.  What do you think about something like `context.advance_cursor(AssetKey("asset_a"))` and `context.advance_all_cursors()` or something similar?  

They might need to work in concert with the event methods (e.g. `context.get_latest_event` / `context.get_all_events`).  For example, let's say that there have been 1000 materializations since the last cursor for a given sensor.  I could see wanting to do some amount of work for every single materialization (maybe there's a backlog), but I am fine with only processing 25 in this sensor evaluation.   Can I just get the earliest 25 materializations (without fetching all from the DB), advance the cursor for just that amount, and wait for the next sensor tick to process the remaining 975 materializations?

I think in order for this to work, `get_all_events` might need to return a generator so that under the hood we  can do batch fetching from the DB.

Reply:
I thought where we ended up was that `return SkipReason(...)` vs `return None` would not be significant for cursor-updating logic, but that we'd punt on adding something like `context.advance_cursor` until later.

Reply:
sandy's correct, the current implementation treats `SkipReason` and `None` the same (and doesn't update the cursor). The idea is to add the base functionality now, and then a follow up PR for allowing users to update the cursor themselves like you mention.

maybe for `advance_cursor` you can pass the asset key and the event record and that will update the cursor to the storage_id in the event record. so then you could process the 25 runs and update the cursor on the last one without needing to know to dig into the event record and get `storage_id`

Reply:
thinking about cursors strikes again.... In the case phil brought up where we have a sensor that is working through a backlog of 1000 asset materializations and launching jobs for 25 materializations per evaluation, we only want to update the cursor to those 25 materializations. But right now if you return any RunRequest the cursor gets updated to the latest materialization. So I think the `advance_cursor` functions need to be included in this PR, and we'll need to make it so that every sensor must manually handle the cursor updating logic.

alternately we could have a bool attribute on the context like `cursor_has_been_updated` that gets flipped to false at the beginning of the evaluation fn. any call to `advance_cursor` flips `cursor_has_been_updated` to true. Then after we have run the user's decorated function, we can check if a run request has been returned AND `cursor_has_been_updated` is false, and if so we can advance the cursor all the way to the latest materializations

Reply:
Mmm - interesting - that makes sense.
 
I don't think it's terrible to expect users to call `advance_cursor` all the time- although I think we should provide some default or other method which makes it really easy to advance the cursor past all the materializations.

If we go this direction, I think maybe we should raise an error if the cursor doesn't get advanced at all, to avoid a that common failure mode.

Also, I like the idea of passing the event record so that users don't need to think about storage IDs.